src/LibcuPDLPx.jl

```diff
@@ -0,0 +1,444 @@
+    vec_dual_prod::Cint
+    ones_primal_d::Ptr{Cdouble}
+    ones_dual_d::Ptr{Cdouble}
+end
+
+struct cupdlpx_result_t
+    num_variables::Cint
+    num_constraints::Cint
+    primal_solution::Ptr{Cdouble}
+    dual_solution::Ptr{Cdouble}
+    total_count::Cint
+    rescaling_time_sec::Cdouble
+    cumulative_time_sec::Cdouble
+    absolute_primal_residual::Cdouble
+    relative_primal_residual::Cdouble
+    absolute_dual_residual::Cdouble
+    relative_dual_residual::Cdouble
+    primal_objective_value::Cdouble
+    dual_objective_value::Cdouble
+    objective_gap::Cdouble
+    relative_objective_gap::Cdouble
+    max_primal_ray_infeasibility::Cdouble
+    max_dual_ray_infeasibility::Cdouble
+    primal_ray_linear_objective::Cdouble
+    dual_ray_objective::Cdouble
+    termination_reason::termination_reason_t
+end
+
+@enum matrix_format_t::UInt32 begin
+    matrix_dense = 0x0000000000000000
+    matrix_csr = 0x0000000000000001
+    matrix_csc = 0x0000000000000002
+    matrix_coo = 0x0000000000000003
+end
+
+struct var"##Ctag#231"
+    data::NTuple{32, UInt8}
+end
+
+function Base.getproperty(x::Ptr{var"##Ctag#231"}, f::Symbol)
+    f === :dense && return Ptr{var"##Ctag#232"}(x + 0)
+    f === :csr && return Ptr{var"##Ctag#233"}(x + 0)
+    f === :csc && return Ptr{var"##Ctag#234"}(x + 0)
+    f === :coo && return Ptr{var"##Ctag#235"}(x + 0)
+    return getfield(x, f)
+end
+
+function Base.getproperty(x::var"##Ctag#231", f::Symbol)
+    r = Ref{var"##Ctag#231"}(x)
+    ptr = Base.unsafe_convert(Ptr{var"##Ctag#231"}, r)
+    fptr = getproperty(ptr, f)
+    GC.@preserve r unsafe_load(fptr)
+end
+
+function Base.setproperty!(x::Ptr{var"##Ctag#231"}, f::Symbol, v)
+    unsafe_store!(getproperty(x, f), v)
+end
+
+function Base.propertynames(x::var"##Ctag#231", private::Bool = false)
+    (:dense, :csr, :csc, :coo, if private
+            fieldnames(typeof(x))
+        else
+            ()
+        end...)
+end
+
+struct matrix_desc_t
+    data::NTuple{56, UInt8}
+end
+
+function Base.getproperty(x::Ptr{matrix_desc_t}, f::Symbol)
+    f === :m && return Ptr{Cint}(x + 0)
+    f === :n && return Ptr{Cint}(x + 4)
+    f === :fmt && return Ptr{matrix_format_t}(x + 8)
+    f === :zero_tolerance && return Ptr{Cdouble}(x + 16)
+    f === :data && return Ptr{var"##Ctag#231"}(x + 24)
+    return getfield(x, f)
+end
+
+function Base.getproperty(x::matrix_desc_t, f::Symbol)
+    r = Ref{matrix_desc_t}(x)
+    ptr = Base.unsafe_convert(Ptr{matrix_desc_t}, r)
+    fptr = getproperty(ptr, f)
+    GC.@preserve r unsafe_load(fptr)
+end
+
+function Base.setproperty!(x::Ptr{matrix_desc_t}, f::Symbol, v)
+    unsafe_store!(getproperty(x, f), v)
+end
+
+function Base.propertynames(x::matrix_desc_t, private::Bool = false)
+    (:m, :n, :fmt, :zero_tolerance, :data, if private
+            fieldnames(typeof(x))
+        else
+            ()
+        end...)
+end
+
+function safe_malloc(size)
+    ccall((:safe_malloc, libcupdlpx), Ptr{Cvoid}, (Csize_t,), size)
+end
```

Comment:
And this should almost certainly not be exposed

Reply:
Got it.